### PR TITLE
feat!: @waku/discovery

### PR DESCRIPTION
**Pull request by danisharora099**
## Problem

We currently have 3 discovery options available within Waku:
- DNS Discovery
- Peer Exchange
- Local Peer Cache (recently added: #1811

For this, we have 3 separate packages, which might only increase with time.

## Solution

Create a generalised package `@waku/discovery` to encapsulate all the available discovery mechanisms.

## Notes

- Resolves https://github.com/waku-org/js-waku/issues/1862

### CODE REVIEW

**Review by adklempner — APPROVED** [review]

File: packages/discovery/package.json [hunk 1]
```diff
@@ -51,6 +51,8 @@
     "node": ">=18"
   },
   "dependencies": {
+    "@waku/interfaces": "0.0.22",
```
Comment (weboko) [inline packages/discovery/package.json:54]: nit: this is `dev` dependency always (except when variables from there are exported)

File: packages/discovery/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// DNS Discovery
+export { PeerDiscoveryDns, wakuDnsDiscovery } from "./dns/dns_discovery.js";
+export { enrTree } from "./dns/constants.js";
+export { DnsNodeDiscovery } from "./dns/dns.js";
+
+// Peer Exchange Discovery
+export {
+  wakuPeerExchange,
```
Comment (weboko) [inline packages/discovery/src/index.ts:8]: Wondering if we need these `waku*`. To me it seems we can avoid and just be straight - `peerExchangeDiscovery`, `peerCacheDiscovery`, etc

File: packages/discovery/src/local-peer-cache/index.spec.ts [hunk 3]
```diff
@@ -11,14 +11,21 @@ import { LocalStoragePeerInfo } from "@waku/interfaces";
...
 
 import { LocalPeerCacheDiscovery } from "./index.js";
 
 chai.use(chaiAsPromised);
 
-global.localStorage = new LocalStorage("./mock_local_storage");
+// dynamically importing the local storage polyfill for node
+if (typeof window === "undefined") {
+  try {
+    const { LocalStorage } = await import("node-localstorage");
```
Comment (weboko) [inline packages/discovery/src/local-peer-cache/index.spec.ts:23]: it's only needed for `tests` so another check can be used too, for example https://github.com/waku-org/js-waku/blob/fb41f4c723238628f398dff4bc423edee8fd1c6e/packages/sdk/src/utils/libp2p.ts#L40

File: release-please-config.json [hunk 4]
```diff
@@ -11,12 +11,10 @@
     "packages/interfaces": {},
     "packages/message-hash": {},
     "packages/enr": {},
-    "packages/peer-exchange": {},
     "packages/core": {},
-    "packages/dns-discovery": {},
     "packages/message-encryption": {},
     "packages/sdk": {},
     "packages/relay": {},
-    "packages/local-peer-cache-discovery": {}
```
Comment (weboko) [inline release-please-config.json:20]: should be removed from here as well - https://github.com/waku-org/js-waku/blob/fb41f4c723238628f398dff4bc423edee8fd1c6e/.release-please-manifest.json#L7

**Review by weboko — CHANGES_REQUESTED** [review]
left some comments

File: packages/discovery/src/index.ts (hunk 2 above)
Comment (danisharora099) [reply packages/discovery/src/index.ts:8]: Removed two exports, but we use reft of them somewhere or the other (mostly for different types of tests)

Please feel free to take a look and suggest if there's an unnecessary export remaining

File: packages/discovery/src/local-peer-cache/index.spec.ts (hunk 3 above)
Comment (danisharora099) [reply packages/discovery/src/local-peer-cache/index.spec.ts:23]: `typeof window` check feels more robust for karma compilation purposes: karma can't compile node packges, thus this acts as a good safe guard.
Is this a nit? can also try with the `NODE_ENV` flag

**Comment by danisharora099** [comment]
> left some comments

Thanks for the comments! Addressed :)

File: packages/discovery/package.json (hunk 1 above)
Comment (danisharora099) [reply packages/discovery/package.json]: https://github.com/waku-org/js-waku/actions/runs/8231612596/job/22507288390?pr=1876

File: .size-limit.cjs [hunk 5]
```diff
@@ -24,7 +24,7 @@ module.exports = [
   },
   {
     name: "DNS discovery",
-    path: "packages/dns-discovery/bundle/index.js",
+    path: "packages/discovery/bundle/index.js",
     import: "{ PeerDiscoveryDns }",
```
Comment (weboko) [inline .size-limit.cjs:28]: let's add other `discovery` entities too

File: packages/discovery/package.json (hunk 1 above)
Comment (weboko) [reply packages/discovery/package.json:54]: ok, let's keep it like this for now and figure it out later https://github.com/waku-org/js-waku/pull/1904#discussion_r1520506480

File: packages/discovery/tsconfig.json [hunk 6]
```diff
@@ -6,5 +6,5 @@
     "tsBuildInfoFile": "dist/.tsbuildinfo"
   },
   "include": ["src"],
-  "exclude": ["src/**/*.spec.ts", "src/test_utils"]
```
Comment (weboko) [inline packages/discovery/tsconfig.json:9]: why is it removed?

**Review by weboko — APPROVED** [review]

File: packages/discovery/tsconfig.json (hunk 6 above)
Comment (danisharora099) [reply packages/discovery/tsconfig.json:9]: `src/test_utils` isn't actually any directory in this package. seems to have come in during a copy. `spec.ts` files are still excluded

**Pull request merged**